src/io_uring/src/lib.rs

```diff
@@ -90,24 +98,48 @@ impl IoUring {
             len if fd < 0 || (len as i32 - 1) < fd => {
                 Err((Error::InvalidFixedFd(fd), op.user_data()))
             }
-            _ => self.squeue.push(unsafe { op.into_sqe() }).map_err(
-                |err_tuple: (SQueueError, T)| -> (Error, T) {
+            _ => self
+                .squeue
+                .push(unsafe { op.into_sqe() })
+                .map(|_| {
+                    self.to_submit = self.to_submit.saturating_add(1);
+                })
+                .map_err(|err_tuple: (SQueueError, T)| -> (Error, T) {
                     (Error::SQueue(err_tuple.0), err_tuple.1)
-                },
-            ),
+                }),
         }
     }
 
     pub fn pop<T>(&mut self) -> Result<Option<Cqe<T>>> {
-        self.cqueue.pop().map_err(Error::CQueue)
+        self.cqueue
+            .pop()
+            .map(|maybe_cqe| {
+                maybe_cqe.map(|cqe| {
+                    self.to_pop = self.to_pop.saturating_sub(1);
+                    cqe
+                })
+            })
+            .map_err(Error::CQueue)
+    }
+
+    fn process_submitted(&mut self, submitted: u32) -> u32 {
+        self.to_submit = 0;
+        self.to_pop = self.to_pop.saturating_add(submitted);
+        submitted
     }
 
     pub fn submit(&mut self) -> Result<u32> {
-        self.squeue.submit(0).map_err(Error::SQueue)
+        self.squeue
+            .submit(self.to_submit, 0)
+            .map(|submitted| self.process_submitted(submitted))
+            .map_err(Error::SQueue)
     }
 
-    pub fn submit_and_wait(&mut self, min_complete: u32) -> Result<u32> {
-        self.squeue.submit(min_complete).map_err(Error::SQueue)
+    pub fn submit_and_wait_all(&mut self) -> Result<u32> {
+        self.squeue
+            .submit(self.to_submit, self.to_submit + self.to_pop)
```

Comment:
Why not check for overflow here too?

Reply:
We shouldn't get an overflow here since `self.to_submit < 32768` and `self.to_pop < 32768 * 2`. I added comments.